[BUG] auditjs guide does not work with forward proxy

**Describe the bug**

We used ossi in the past which works fine with a forward proxy.
The proxy is configured with `http_proxy` and `https_proxy` environment variables.
We switched to guide (because ossi is deprecated) and auditjs now fails with: 

```[2026-06-01T07:03:02.078] [ERROR] auditjs - There was an error auditing with Sonatype Guide Error: There was an error making the request to Sonatype Guide: FetchError: The request failed and the interceptors did not return an alternative response
 at GuideRequestService.getResultsFromGuide (/home/node/.npm/_npx/db34456ff7bcaff8/node_modules/auditjs/bin/Services/GuideRequestService.js:110:19)
 at process.processTicksAndRejections (node:internal/process/task_queues:103:5)
 at async Promise.all (index 0)
 at async Application.auditWithGuide (/home/node/.npm/_npx/db34456ff7bcaff8/node_modules/auditjs/bin/Application/Application.js:216:25)
 at async Application.startApplication (/home/node/.npm/_npx/db34456ff7bcaff8/node_modules/auditjs/bin/Application/Application.js:96:13)```

We no longer see a proxy connect in our proxy logs. 
We investigated the code and OssIndexRequestService creates a ProxyAgent on line: https://github.com/sonatype-nexus-community/auditjs/blob/01a77914983c098e42412378da912b0dcda1413b/src/Services/OssIndexRequestService.ts#L66

The new guide GuideRequestService uses the OSSIndexCompatibilityApi from dependency @sonatype/sonatype-guide-api-client which does not seem to configure a HttpAgent with proxy support as dispatcher.

Could support for a forward proxy while using guide be provided?
This currently blocks us from using auditjs with guide as a forward proxy is required in our environment.

**To Reproduce**

Use auditjs guide with a forward proxy

**Expected behavior**

Successful sonatype guide API requests.

**Desktop (please complete the following information):**
 - OS: linux
 - NodeJS Version: 22
 - Version 5.0.1

cc @sonatype-nexus-community/auditjs

--- a/package.json
+++ b/package.json
@@ -85,7 +85,7 @@
     "@xmldom/xmldom": "^0.9.0",
     "chalk": "^4.1.2",
     "figlet": "^1.2.4",
-    "js-yaml": "^4.1.0",
+    "js-yaml": "^4.2.0",
     "log4js": "^6.4.0",
     "node-persist": "^3.1.0",
     "ora": "^5.4.1",

--- a/src/Services/GuideRequestService.spec.ts
+++ b/src/Services/GuideRequestService.spec.ts
@@ -49,6 +49,7 @@ describe('GuideRequestService', () => {
     vi.unstubAllGlobals();
     delete process.env.http_proxy;
     delete process.env.https_proxy;
+    mockFetch.mockClear();
   });
 
   it('sends Authorization: Bearer when accessToken is set (PAT token mode)', async () => {
@@ -161,7 +162,9 @@ describe('GuideRequestService', () => {
     });
 
     it('should not include dispatcher when no proxy is configured', async () => {
-      // No proxy env vars set
+      // Ensure no proxy env vars are set before creating service
+      delete process.env.http_proxy;
+      delete process.env.https_proxy;
 
       const expectedOutput = [
         {
@@ -186,7 +189,8 @@ describe('GuideRequestService', () => {
       const fetchCall = mockFetch.mock.calls[0];
       const fetchOptions = fetchCall[1] as RequestInit & { dispatcher?: unknown };
 
-      // When no proxy is configured, dispatcher should be undefined
+      // When no proxy is configured, the custom fetchApi should not be provided,
+      // so dispatcher should not be in the options
       expect(fetchOptions.dispatcher).toBeUndefined();
     });
   });

--- a/src/Services/GuideRequestService.spec.ts
+++ b/src/Services/GuideRequestService.spec.ts
@@ -129,23 +129,22 @@ describe('GuideRequestService', () => {
   });
 
   describe('proxy support', () => {
-    it('should pass ProxyAgent as dispatcher when http_proxy is set', async () => {
-      process.env.http_proxy = 'http://proxy.example.com:8080';
-
-      const expectedOutput = [
+    const mockSuccessResponse = {
+      ok: true,
+      status: 200,
+      statusText: 'OK',
+      json: vi.fn().mockResolvedValue([
         {
           coordinates: 'pkg:npm/test@1.0.0',
           reference: 'https://guide.sonatype.com/blah',
           vulnerabilities: [],
         },
-      ];
+      ]),
+    };
 
-      mockFetch.mockResolvedValueOnce({
-        ok: true,
-        status: 200,
-        statusText: 'OK',
-        json: vi.fn().mockResolvedValue(expectedOutput),
-      });
+    it('should pass ProxyAgent as dispatcher when http_proxy is set', async () => {
+      process.env.http_proxy = 'http://proxy.example.com:8080';
+      mockFetch.mockResolvedValueOnce(mockSuccessResponse);
 
       const svc = new GuideRequestService('user', 'token', CACHE_PATH, SERVER);
       const coords = [new Coordinates('test', '1.0.0')];
@@ -153,10 +152,7 @@ describe('GuideRequestService', () => {
 
       // Verify fetch was called with a dispatcher (ProxyAgent)
       expect(mockFetch).toHaveBeenCalled();
-      const fetchCall = mockFetch.mock.calls[0];
-      const fetchOptions = fetchCall[1] as RequestInit & { dispatcher?: ProxyAgent };
-
-      // This assertion demonstrates the bug - dispatcher should be a ProxyAgent but is undefined
+      const fetchOptions = mockFetch.mock.calls[0][1] as RequestInit & { dispatcher?: ProxyAgent };
       expect(fetchOptions.dispatcher).toBeDefined();
       expect(fetchOptions.dispatcher).toBeInstanceOf(ProxyAgent);
     });
@@ -165,32 +161,14 @@ describe('GuideRequestService', () => {
       // Ensure no proxy env vars are set before creating service
       delete process.env.http_proxy;
       delete process.env.https_proxy;
-
-      const expectedOutput = [
-        {
-          coordinates: 'pkg:npm/test@1.0.0',
-          reference: 'https://guide.sonatype.com/blah',
-          vulnerabilities: [],
-        },
-      ];
-
-      mockFetch.mockResolvedValueOnce({
-        ok: true,
-        status: 200,
-        statusText: 'OK',
-        json: vi.fn().mockResolvedValue(expectedOutput),
-      });
+      mockFetch.mockResolvedValueOnce(mockSuccessResponse);
 
       const svc = new GuideRequestService('user', 'token', CACHE_PATH, SERVER);
       const coords = [new Coordinates('test', '1.0.0')];
       await svc.callGuideOrGetFromCache(coords, 'npm');
 
       expect(mockFetch).toHaveBeenCalled();
-      const fetchCall = mockFetch.mock.calls[0];
-      const fetchOptions = fetchCall[1] as RequestInit & { dispatcher?: unknown };
-
-      // When no proxy is configured, the custom fetchApi should not be provided,
-      // so dispatcher should not be in the options
+      const fetchOptions = mockFetch.mock.calls[0][1] as RequestInit & { dispatcher?: unknown };
       expect(fetchOptions.dispatcher).toBeUndefined();
     });
   });

--- a/src/Services/GuideRequestService.spec.ts
+++ b/src/Services/GuideRequestService.spec.ts
@@ -20,6 +20,7 @@ import { OSSIndexCompatibilityApi } from '@sonatype/sonatype-guide-api-client';
 import type { InitOverrideFunction } from '@sonatype/sonatype-guide-api-client';
 import { Coordinates } from '../Types/Coordinates';
 import { rmSync, existsSync } from 'node:fs';
+import { ProxyAgent } from 'undici';
 
 // node-persist is mocked globally so tests never touch the filesystem cache.
 vi.mock('node-persist', () => ({
@@ -36,12 +37,18 @@ const CACHE_PATH = '/tmp/.sonatype-guide-test';
 const SERVER = 'https://api.guide.sonatype.com';
 
 describe('GuideRequestService', () => {
+  const mockFetch = vi.fn();
+
   beforeEach(() => {
     if (existsSync(CACHE_PATH)) rmSync(CACHE_PATH, { recursive: true, force: true });
+    vi.stubGlobal('fetch', mockFetch);
   });
 
   afterEach(() => {
     vi.restoreAllMocks();
+    vi.unstubAllGlobals();
+    delete process.env.http_proxy;
+    delete process.env.https_proxy;
   });
 
   it('sends Authorization: Bearer when accessToken is set (PAT token mode)', async () => {
@@ -117,6 +124,70 @@ describe('GuideRequestService', () => {
       // No HTTP request should be made; the result is an empty array rather than an error.
       expect(getResultsSpy).not.toHaveBeenCalled();
       expect(result).toEqual([]);
+    });
+  });
+
+  describe('proxy support', () => {
+    it('should pass ProxyAgent as dispatcher when http_proxy is set', async () => {
+      process.env.http_proxy = 'http://proxy.example.com:8080';
+
+      const expectedOutput = [
+        {
+          coordinates: 'pkg:npm/test@1.0.0',
+          reference: 'https://guide.sonatype.com/blah',
+          vulnerabilities: [],
+        },
+      ];
+
+      mockFetch.mockResolvedValueOnce({
+        ok: true,
+        status: 200,
+        statusText: 'OK',
+        json: vi.fn().mockResolvedValue(expectedOutput),
+      });
+
+      const svc = new GuideRequestService('user', 'token', CACHE_PATH, SERVER);
+      const coords = [new Coordinates('test', '1.0.0')];
+      await svc.callGuideOrGetFromCache(coords, 'npm');
+
+      // Verify fetch was called with a dispatcher (ProxyAgent)
+      expect(mockFetch).toHaveBeenCalled();
+      const fetchCall = mockFetch.mock.calls[0];
+      const fetchOptions = fetchCall[1] as RequestInit & { dispatcher?: ProxyAgent };
+
+      // This assertion demonstrates the bug - dispatcher should be a ProxyAgent but is undefined
+      expect(fetchOptions.dispatcher).toBeDefined();
+      expect(fetchOptions.dispatcher).toBeInstanceOf(ProxyAgent);
+    });
+
+    it('should not include dispatcher when no proxy is configured', async () => {
+      // No proxy env vars set
+
+      const expectedOutput = [
+        {
+          coordinates: 'pkg:npm/test@1.0.0',
+          reference: 'https://guide.sonatype.com/blah',
+          vulnerabilities: [],
+        },
+      ];
+
+      mockFetch.mockResolvedValueOnce({
+        ok: true,
+        status: 200,
+        statusText: 'OK',
+        json: vi.fn().mockResolvedValue(expectedOutput),
+      });
+
+      const svc = new GuideRequestService('user', 'token', CACHE_PATH, SERVER);
+      const coords = [new Coordinates('test', '1.0.0')];
+      await svc.callGuideOrGetFromCache(coords, 'npm');
+
+      expect(mockFetch).toHaveBeenCalled();
+      const fetchCall = mockFetch.mock.calls[0];
+      const fetchOptions = fetchCall[1] as RequestInit & { dispatcher?: unknown };
+
+      // When no proxy is configured, dispatcher should be undefined
+      expect(fetchOptions.dispatcher).toBeUndefined();
     });
   });
 });

--- a/src/Services/GuideRequestService.ts
+++ b/src/Services/GuideRequestService.ts
@@ -15,12 +15,13 @@
  */
 
 import { OSSIndexCompatibilityApi, RecommendationsApi, Configuration } from '@sonatype/sonatype-guide-api-client';
-import type { RecommendationResponse, InitOverrideFunction } from '@sonatype/sonatype-guide-api-client';
+import type { RecommendationResponse, InitOverrideFunction, FetchAPI } from '@sonatype/sonatype-guide-api-client';
 import NodePersist from 'node-persist';
 import path from 'path';
 import { homedir } from 'os';
 import { Coordinates } from '../Types/Coordinates';
 import { OssIndexServerResultJSON } from '../Types/OssIndexServerResult';
+import { RequestHelpers } from './RequestHelpers';
 
 const GUIDE_BASE_URL = 'https://api.guide.sonatype.com';
 
@@ -41,20 +42,27 @@ export class GuideRequestService {
     readonly server: string = GUIDE_BASE_URL,
     readonly accessToken?: string,
   ) {
+    // Create a custom fetch wrapper that includes proxy support via dispatcher
+    const proxyAgent = RequestHelpers.getHttpAgent();
+    const fetchApi: FetchAPI | undefined = proxyAgent
+      ? (url: string | URL | Request, init?: RequestInit) =>
+          fetch(url, { ...init, dispatcher: proxyAgent })
+      : undefined;
+
     // OSSIndexCompatibilityApi only supports HTTP Basic auth.
     // In PAT-only mode (no username), send the PAT as password with empty username
     // so the generated client includes an Authorization: Basic :<PAT> header.
     const ossUsername = username ?? (accessToken ? '' : undefined);
     const ossPassword = token ?? accessToken;
     this.api = new OSSIndexCompatibilityApi(
-      new Configuration({ username: ossUsername, password: ossPassword, basePath: server }),
+      new Configuration({ username: ossUsername, password: ossPassword, basePath: server, fetchApi }),
     );
 
     // RecommendationsApi supports Bearer auth; fall back to Basic when username is present.
     const recConfig =
       accessToken && !username
-        ? new Configuration({ accessToken, basePath: server })
-        : new Configuration({ username, password: token, basePath: server });
+        ? new Configuration({ accessToken, basePath: server, fetchApi })
+        : new Configuration({ username, password: token, basePath: server, fetchApi });
     this.recommendationsApi = new RecommendationsApi(recConfig);
   }
 

--- a/src/Services/GuideRequestService.ts
+++ b/src/Services/GuideRequestService.ts
@@ -46,7 +46,7 @@ export class GuideRequestService {
     const proxyAgent = RequestHelpers.getHttpAgent();
     const fetchApi: FetchAPI | undefined = proxyAgent
       ? (url: string | URL | Request, init?: RequestInit) =>
-          fetch(url, { ...init, dispatcher: proxyAgent })
+          fetch(url, { ...init, dispatcher: proxyAgent } as unknown as RequestInit)
       : undefined;
 
     // OSSIndexCompatibilityApi only supports HTTP Basic auth.

--- a/src/Services/OssIndexRequestService.spec.ts
+++ b/src/Services/OssIndexRequestService.spec.ts
@@ -18,6 +18,7 @@ import { expect, vi, describe, it, afterEach, beforeEach } from 'vitest';
 import { OssIndexRequestService } from './OssIndexRequestService';
 import { Coordinates } from '../Types/Coordinates';
 import { rmSync, existsSync } from 'fs';
+import { ProxyAgent } from 'undici';
 
 // This will only work on Linux/OS X; find a better Windows-friendly path
 const CACHE_LOCATION = '/tmp/.ossindex';
@@ -34,6 +35,8 @@ describe('OssIndexRequestService', () => {
   afterEach(() => {
     vi.clearAllMocks();
     vi.unstubAllGlobals();
+    delete process.env.http_proxy;
+    delete process.env.https_proxy;
   });
 
   it('should have its request rejected when the OSS Index server is down', async () => {
@@ -65,5 +68,35 @@ describe('OssIndexRequestService', () => {
     const coords = [new Coordinates('commander', '2.12.2', '@types')];
     const result = await requestService.callOSSIndexOrGetFromCache(coords);
     expect(result).toEqual(expectedOutput);
+  });
+
+  it('should pass ProxyAgent as dispatcher when http_proxy is set', async () => {
+    if (existsSync(CACHE_LOCATION)) rmSync(CACHE_LOCATION, { recursive: true, force: true });
+    process.env.http_proxy = 'http://proxy.example.com:8080';
+
+    const expectedOutput = [
+      {
+        coordinates: 'pkg:npm/test@1.0.0',
+        reference: 'https://ossindex.sonatype.org/blah',
+        vulnerabilities: [],
+      },
+    ];
+
+    mockFetch.mockResolvedValueOnce({
+      ok: true,
+      statusText: 'OK',
+      json: vi.fn().mockResolvedValue(expectedOutput),
+    });
+
+    const requestService = new OssIndexRequestService(undefined, undefined, CACHE_LOCATION, OSS_INDEX_BASE_URL);
+    const coords = [new Coordinates('test', '1.0.0')];
+    await requestService.callOSSIndexOrGetFromCache(coords);
+
+    // Verify fetch was called with a dispatcher (ProxyAgent)
+    expect(mockFetch).toHaveBeenCalled();
+    const fetchCall = mockFetch.mock.calls[0];
+    const fetchOptions = fetchCall[1] as RequestInit & { dispatcher?: ProxyAgent };
+    expect(fetchOptions.dispatcher).toBeDefined();
+    expect(fetchOptions.dispatcher).toBeInstanceOf(ProxyAgent);
   });
 });

